kt/godot-runtime/src/main/kotlin/godot/core/GarbageCollector.kt

```diff
@@ -0,0 +1,143 @@
+import java.time.Instant
+
+object GarbageCollector : Thread() {
+    val refWrappedMap = mutableMapOf<VoidPtr, WeakReference<KtObject>>()
+    val wrappedMap = mutableMapOf<VoidPtr, KtObject>()
+    val refWrappedSuppressBuffer = mutableListOf<VoidPtr>()
+    val wrappedSuppressBuffer = mutableListOf<VoidPtr>()
+
+    private var forceJvmGarbageCollector = false
+    private var gcState = GCState.NONE
+    val isClosed: Boolean
+        get() = gcState == GCState.CLOSED
+
+    fun registerInstance(instance: KtObject) {
+        synchronized(this) {
+            val rawPtr = instance.rawPtr
+            if (instance.isRef) {
+                refWrappedMap[rawPtr] = WeakReference(instance)
+                MemoryBridge.ref(rawPtr)
+            } else {
+                wrappedMap[rawPtr] = instance
+            }
+        }
+    }
+
+    fun getObjectInstance(ptr: VoidPtr): KtObject? {
+        val ktObject = wrappedMap[ptr]
+        return if (ktObject != null) {
+            if (MemoryBridge.checkInstance(ptr, ktObject.godotInstanceId)) ktObject else null
+        } else null
+    }
+
+    fun start(forceJvmGarbageCollector: Boolean) {
+        this.forceJvmGarbageCollector = forceJvmGarbageCollector
+        start()
+        gcState = GCState.STARTED
+    }
+
+    fun close() {
+        gcState = GCState.CLOSED
+        println("Closing GC thread ...")
+    }
+
+    fun cleanUp() {
+        val begin = Instant.now()
+        while (refWrappedMap.isNotEmpty() || wrappedMap.isNotEmpty()) {
+            forceJvmGc()
+            checkAndClean()
+            val finish = Instant.now()
+            if (Duration.between(begin, finish).toMillis() > 5000) {
+                throw GCEndException(
+                        buildString {
+                            append("Some JVM godot instances are leaked.")
+                            append(System.lineSeparator())
+                            append("Leaked references:")
+                            append(System.lineSeparator())
+                            for (entry in refWrappedMap) {
+                                append("    ${entry.key}: ${entry.value}")
+                                append(System.lineSeparator())
+                            }
+                            append("Leaked objects:")
+                            append(System.lineSeparator())
+                            for (entry in wrappedMap) {
+                                append("    ${entry.key}: ${entry.value}")
+                                append(System.lineSeparator())
+                            }
+                        }
+                )
+            }
+        }
+
+    }
+
+    override fun run() {
+        while (gcState == GCState.STARTED) {
+            if (forceJvmGarbageCollector) {
+                forceJvmGc()
+            }
+            checkAndClean()
+            sleep(500)
+        }
+        gcState = GCState.CLOSED
+    }
+
+    private fun checkAndClean() {

```

Comment:
We probably need to synchronize this with `registerInstance`. What if we move the state of the GC locally, i.e:

```kt
fun checkAndClean() {
	lateinit var refWrappedMapCopy: ...;
    // ...
	synchronize(this) {
         val refWrappedMapCopy = refWrappedMap;
         // do the same as the others

         refWrappedMap = new ...;
    }

    // do gc checks with the copy
}
```

Reply:
ahh, we still need to support `getObjectInstance` while running the gc.

Reply:
Making a copy of the container seems heavy for real life applications.
We should probably prevent direct access to the maps and create methods for external access.
Code outside of the GC only need to read and we only be cautious about thread safety when writing, the only place where it's the case is when registering/removing instead bur I'm going to add queues so they are done in the GC thread.
We don't need to lock when we do the checks themselves because in that location we know we are in the GC thread so writing is not possible at that moment (because only the GC thread will write).